Add reboot action plugin

samdoran opened the pull request:
##### SUMMARY

Create a `reboot` action plugin for rebooting *nix hosts.

Refactor `win_reboot` so it subclasses methods in `reboot.py`.

Fixes #16186

##### ISSUE TYPE

 - New Module Pull Request


##### COMPONENT NAME
`reboot.py`

##### ANSIBLE VERSION
<!--- Paste verbatim output from "ansible --version" between quotes below -->
```
2.7
```

dagwieers commented:
Hmm, you know that I already have a semi-working implementation ?
https://github.com/dagwieers/ansible/commit/a69dbd1abb1710cc346b60831f19159a56102053

If this was important to hit v2.5 I could have finished it.

samdoran commented:
@dagwieers I did see that. I wanted to got with a separate module for *nix rather than one that does everything.

dagwieers commented:
@samdoran Supporting Windows is not the hardest part though. You almost get that for free.

samdoran commented:
@dagwieers I’m back at work next week. Let’s work on getting this feature implemented and m erged. I don’t care about using this code, I just want this functionality to be in Ansible as you highlight in this issue.

samdoran committed:
Update docs

samdoran committed:
Add reboot action plugin

Refactor win_reboot so it is subclassed from reboot

samdoran committed:
Use new connection methods

samdoran committed:
Test fixes

samdoran committed:
Use better uptime command for Linux

Use who -b to get the last time the system was booted rather than uptime, which changes every second.

samdoran committed:
Use distribution specefic commands and flags

Query the managed node to determien its distribution, then set the appropriate command and flags.

samdoran committed:
Tune debug messages a bit

samdoran committed:
Update module docs with details about pre_reboot_delay

s docs

samdoran committed:
Ensure that post_reboot_delay is a positive number

samdoran committed:
Remove the stringification

samdoran committed:
Add integration tests

samdoran committed:
Make sure aliases are honored

abadger commented on lib/ansible/modules/system/reboot.py:
```diff
@@ -0,0 +1,82 @@
+#!/usr/bin/python
+# -*- coding: utf-8 -*-
+
```
Copyright line

abadger commented on lib/ansible/plugins/action/reboot.py:
```diff
@@ -0,0 +1,260 @@
+        except Exception as e:
+            result['failed'] = True
+            result['reboot'] = False
+            result['msg'] = to_native(e)
```
We're in the controller and this isn't something we're passing to an exception constructor.  Therefore use to_text() here.

samdoran committed:
Handle systems that have an incorrect last boot time

SystemD and fakehw-clock do not properly set the
last boot time and instead always set it to epoch.
Use a different command if that is the case.

samdoran committed:
Copyright and encoding fixes

webknjaz commented on lib/ansible/plugins/action/reboot.py:
```diff
@@ -0,0 +1,266 @@
+        # people will lose waiting for that extra 1 minute delay and want to give them their
+        # lives back.
+        delay_min = pre_reboot_delay // 60
+        delay_min_macos = pre_reboot_delay // 60 | 1
```
`delay_min_macos = delay_min | 1`

webknjaz commented on lib/ansible/plugins/action/reboot.py:
```diff
@@ -0,0 +1,266 @@
+    def do_until_success_or_timeout(self, action, reboot_timeout, before_uptime, action_desc, fail_sleep=1):
+        max_end_time = datetime.utcnow() + timedelta(seconds=reboot_timeout)
+
+        while datetime.utcnow() < max_end_time:
```
Maybe implement exponential backoff?

samdoran replied:
I'm not sure how to do that. Would love to learn!

samdoran replied:
I have something working. I'll push it up for you to review shortly. Thanks for the tip!

webknjaz replied:
I like it!

samdoran replied:
Thanks!

webknjaz commented on lib/ansible/plugins/action/win_reboot.py:
```diff
@@ -1,16 +1,15 @@
-# (c) 2016, Matt Davis <mdavis@ansible.com>
+# (c) 2018, Matt Davis <mdavis@ansible.com>
```
Why change year? Why not 2016-2018 then?

samdoran replied:
I just figured copyright dates were supposed to be updated to the current year. I don't pretend to understand the law around the dates.

webknjaz replied:
I'm not sure. Normally, the year in the year of content creation (first publication) and 2016 means "copyright protected 2016 and forward".
If there's content created during a period of time it'd be a range or a list of years.

https://smallbusiness.chron.com/year-copyright-statement-mean-62027.html
https://www.copyrightlaws.com/copyright-notice-year/

webknjaz replied:
Also, there was some guideline where the next updated would add a separate line with themselves, but won't edit other lines.

cc @abadger

webknjaz commented on lib/ansible/modules/system/reboot.py:
```diff
@@ -0,0 +1,82 @@
+
+DOCUMENTATION = r'''
+module: reboot
+short_description: Reboot a \*nix machine
```
I believe you don't need `\*`, just `*` in raw-string literal

webknjaz replied:
Oh, is that for YAML? It shouldn't need that as well except for the beginning of the value.

samdoran replied:
The only way I could get the docs sanity test to pass was escaping the `*`. I figured having the documentation as a raw string would suffice, but it still failed sanity tests.

samdoran replied:
It's to make the `rst` parser happy. It complains about an unbalanced emphasis character.

dagwieers replied:
I would not mention *nix in the short_description, as we want to make this support Windows as well.

sivel replied:
There is a separate `win_reboot` module.

dagwieers replied:
@sivel Sure, but since reboot supports multiple platforms anyway, we might as well make it work on Windows too. That was the original intention anyway, and it helps doing reboots without having to differentiate between platform (i.e. doing them consecutively).

sivel replied:
That is not the intention of this PR.  In this PR there are 2 separate action plugins/modules, and the `win_reboot` inherits from `reboot` and overrides for it's specific functionality.  `reboot` only supports nix machines.

I'm not the author here of course, so I'll let @samdoran address this further.

dagwieers replied:
@sivel I was talking about the original issue and example code. And I am not saying we need this as part of this PR, but if the intention was to add it, we might as well not say *nix in the short_description and provide details in the description.

PS We don't mention *nix in the **user** or **service** module either.

dagwieers replied:
In other modules, we add a reference in the documentation to the Windows counterpart (and vice-versa). We should do this here as well (although I would prefer Windows-support in this module at some point ;-))

But again, I don't think we need to add *nix to the short_description.

nitzmahone replied:
This was parameter-compatible with `win_reboot` at the moment, but I suspect it may not always remain so. For instance, I've had a feature I've wanted to add to win_reboot that optionally checks all the various "do I need a reboot?" flags which would be nonsensical for non-Windows platforms- it's kind of the same reason we split `win_copy` from `copy`... Trying to force these things to be arg/doc compatible will either force lowest-common-denominator behavior or have an increasing number of "this parameter only applies to platform X" kinds of things. It's a tough call, but I'd rather keep them as separate actions with shared impl where appropriate.

dagwieers replied:
@nitzmahone Same for Solaris, AIX, FreeBSD, Linux, etc. We already have to provide guidance on how parameters influences each. If you're managing systems in a heterogeneous environment, being able to reboot at the same time is very useful. Having to wait for Windows systems to be reboot first, before you can reboot the Linux systems is just annoying. (And I bet not a lot of people manage heterogeneous environments, so I will be isolated on this one again)

samdoran replied:
The intent was to use `reboot` as a base class that `win_reboot` could subclass since some of the reboot logic needs to be quite a bit different for Windows. Yes, it's not great that you have to do separate tasks to reboot non-homogenous environments, but I wanted to maintain the separation between the two modules to allow them to diverge as needed.

I will remove the reference to `*nix` in the documentation.

samdoran committed:
Minor fixes based on feedback

samdoran committed:
Add exponential backoff to sucess check method

samdoran committed:
Update integration test

Skip the integration test if it would try to reboot the control node. We need a new mechanism to account for this scenario in ansible-test, so tests must currently be run manually for this plugin.

dagwieers commented:
How do we know that the reboot was successful ?
Checking for uptime could be useful here (see #16186)

nitzmahone commented on lib/ansible/modules/system/reboot.py:
```diff
@@ -0,0 +1,82 @@
+    type: int
+  connect_timeout:
+    description:
+      - Maximum seconds to wait for a single successful TCP connection to the managed hosts before trying again.
```
s/single successful TCP/successful/

nitzmahone commented on lib/ansible/modules/system/reboot.py:
```diff
@@ -0,0 +1,82 @@
+    description:
+      - Maximum seconds to wait for a single successful TCP connection to the managed hosts before trying again.
+      - This defaults to None which means the default setting for the underlying connection plugin is used.
+    default: None
```
Probably just get rid of the `default` line, since this is the default for everything- change the text above to something like "If unspecified, the underlying connection plugin's timeout is used"

nitzmahone commented on lib/ansible/plugins/action/win_reboot.py:
```diff
@@ -23,201 +22,68 @@ class TimedOutException(Exception):
+
         # Get the original connection_timeout option var so it can be reset after
-        connection_timeout_orig = None
+        result['connection_timeout_orig'] = None
```
You're using the result object to smuggle this around, but it should probably be an instance field instead, since we don't actually want it in the results. Probably the same for the other cases where we're using the result dict as intermediate storage- put them in fields and just construct the result object at the end before it's returned.

nitzmahone commented on lib/ansible/plugins/action/reboot.py:
```diff
@@ -0,0 +1,277 @@
+    def run(self, tmp=None, task_vars=None):
+        self._supports_check_mode = True
+        self._supports_async = True
+
```
Probably should also check and fail for `self._connection.transport == 'local'` - no need to go nutty trying to figure out if they're rebooting the controller via a more circuitous route, but at least cover the easy one.

nitzmahone reviewed (changes requested):
a few minor things, but pretty close

samdoran commented:
@dagwieers 
> How do we know that the reboot was successful ?

`get_system_uptime()` runs `who -b`, which reports the last time the system was booted. The plugin captures that before performing a reboot, then continually checks for that value to change until it hits a timeout.

On systems that lack a real time clock (RTC), it falls back to using `uptime -s` to determine if the system has rebooted.

I found that looking at the `uptime` string was more difficult since the format varies between platforms, and it changes every second.

Those were the most universal commands I could find. The `-s` flag is pretty inconsistent in `uptime`, but on systems that lack a RTC (usually Rasperry Pi-like devices), they seem to have the `-s` flag.

Not perfect, but it covers most of the bases.

samdoran committed:
Fail early with running with local connection

samdoran committed:
Update docs based on feedback

samdoran committed:
Update integration test

Skip the integration test if it would try to reboot the control node. We need a new mechanism to account for this scenario in ansible-test, so tests must currently be run manually for this plugin.

nitzmahone commented on lib/ansible/plugins/action/reboot.py:
```diff
@@ -0,0 +1,277 @@
+    }
+
+    def deprecated_args(self):
+        deprecated_args = self._task.args.get('DEFAULT_DEPRECATED_ARGS')
```
I think this is busted, but will file a bug to fix it after

nitzmahone commented on lib/ansible/plugins/action/reboot.py:
```diff
@@ -0,0 +1,277 @@
+    }
+
+    def deprecated_args(self):
+        deprecated_args = self._task.args.get('DEFAULT_DEPRECATED_ARGS')
```
shouldn't this line and the one below just not be here, and the loop be `for arg, version in self.DEFAULT_DEPRECATED_ARGS` (though those aren't really a default either, so `DEFAULT` is a bit of a misnomer)

samdoran replied:
Yup.

samdoran replied:
I made the changes you suggested.

nitzmahone committed:
minor refactoring, state mgmt changes

nitzmahone commented:
(unrelated integration test failure)
